### PR TITLE
CR-3568: Parsing evidence, added tests, improved performance

### PR DESCRIPTION
**Pull request by Tomerva**

### CODE REVIEW

File: src/sidebar/cloudrail_sidebar_provider.ts [hunk 1]
```diff
@@ -107,4 +113,20 @@ export class CloudrailSidebarProvider implements vscode.TreeDataProvider<Cloudra
         }
         return ruleTreeItem;
     } 
+
+    private parseEvidence(evidence: string): string {
```
Comment (Tomerva) [inline src/sidebar/cloudrail_sidebar_provider.ts:117]: This function will should be changed according to specification -
what to do with: ~, `

File: src/commands/scan.ts [hunk 2]
```diff
@@ -66,59 +59,27 @@ export async function scan(diagnostics: vscode.DiagnosticCollection, sidebarProv
...
 
             progress.report({ increment: 100, message: 'Applying scan results...'});
-            await handleRunResults(runResults!, diagnostics, terraformWorkingDirectory, sidebarProvider);
+            runResultPublisher.updateRunResults(runResults, terraformWorkingDirectory);
             isSuccessfulScan = true;
         });
     } catch(e) {
         logger.error(`Failed to perform scan. reason: ${e}`);
-        vscode.window.showErrorMessage(`An unknown error has occured while performing the scan. Check log for more information: ${logPath}`);
+        if (!await CloudrailRunner.getCloudrailVersion()) {
```
Comment (alonashkenazi) [inline src/commands/scan.ts:67]: If the installation will fail, will be in an infinite loop?

File: src/tools/run_result_diagnostic_subscriber.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+export default class RunResultDiagnosticSubscriber implements RunResultsSubscriber{
+    constructor(private diagnostics: vscode.DiagnosticCollection) {}
+
+    async assessmentStart(): Promise<void> {
+        this.diagnostics.clear();
+    }
+
+    async assessmentFailed(): Promise<void> {}
+
+    async updateRunResults(runResults: CloudrailRunResponse, ruleResults: RuleResult[], terraformWorkingDirectory: string): Promise<void> {
```
Comment (alonashkenazi) [inline src/tools/run_result_diagnostic_subscriber.ts:17]: think of separating the logic that converts the Response to diagnostic to a different function so you will be able to create a unit test for the conversion (same for CloudrailSidebarProvider.updateRunResults)

File: src/tools/run_result_diagnostic_subscriber.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import { EvidenceFormat, parseEvidence } from './parse_utils';
+
+export default class RunResultDiagnosticSubscriber implements RunResultsSubscriber{
+    constructor(private diagnostics: vscode.DiagnosticCollection) {}
+
+    async assessmentStart(): Promise<void> {
+        this.diagnostics.clear();
+    }
+
+    async assessmentFailed(): Promise<void> {}
```
Comment (alonashkenazi) [inline src/tools/run_result_diagnostic_subscriber.ts:15]: in the other subscriber you "clear" the results by changing the message to "Last scan failed" why here you are not clearing the results?

File: src/tools/parse_utils.ts [hunk 5]
```diff
@@ -4,4 +4,28 @@ import util from 'util';
...
+
+export enum EvidenceFormat {
+    plainText = 0,
+    html = 1
+}
+
+export function parseEvidence(evidence: string, format: EvidenceFormat): string {
+    let parsedEvidence = format === EvidenceFormat.html ? parseHtmlLinks(evidence) : evidence.replace(/</g, '').replace(/>/g, '');
+
+    parsedEvidence = parsedEvidence.replace(/`.*`/g, (code) => {
```
Comment (alonashkenazi) [inline src/tools/parse_utils.ts:24]: use const with appropriate names for the special characters
export const EMPHASIZE = '`'; // code block
export const INDENT = '~';
export const SENTENCE = '. ';
export const LINK = '<';
export const CLOSING_LINK = '>';

File: src/run_result_subscriber.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { CloudrailRunResponse } from './cloudrail_runner';
```
Comment (alonashkenazi) [inline src/run_result_subscriber.ts:1]: move to a subscribers folder?

File: src/sidebar/cloudrail_sidebar_provider.ts [hunk 7]
```diff
@@ -1,10 +1,13 @@
 import path from 'path';
```
Comment (alonashkenazi) [inline src/sidebar/cloudrail_sidebar_provider.ts:1]: move to subscriber/sidebar/ folder?

**Review by alonashkenazi — APPROVED** [review]

File: src/tools/run_result_diagnostic_subscriber.ts (hunk 4 above)
Comment (Tomerva) [reply src/tools/run_result_diagnostic_subscriber.ts:15]: already cleared by the assessmentStart()

**Pull request merged**